plugins/quota-aware-fs/src/main/java/org/elasticsearch/fs/quotaaware/QuotaAwareFileSystemProvider.java

```diff
@@ -0,0 +1,464 @@
+    }
+
+    @Override
+    public <V extends FileAttributeView> V getFileAttributeView(Path path, Class<V> type, LinkOption... options) {
+        return delegate.getFileAttributeView(QuotaAwarePath.unwrap(path), type, options);
+    }
+
+    @Override
+    public <A extends BasicFileAttributes> A readAttributes(Path path, Class<A> type, LinkOption... options) throws IOException {
+        try {
+            return delegate.readAttributes(QuotaAwarePath.unwrap(path), type, options);
+        } catch (ProviderMismatchException e) {
+            throw new IllegalArgumentException("Failed to read attributes for path: [" + path + "]", e);
+        }
+    }
+
+    @Override
+    public Map<String, Object> readAttributes(Path path, String attributes, LinkOption... options) throws IOException {
+        return delegate.readAttributes(QuotaAwarePath.unwrap(path), attributes, options);
+    }
+
+    @Override
+    public void setAttribute(Path path, String attribute, Object value, LinkOption... options) throws IOException {
+        delegate.setAttribute(QuotaAwarePath.unwrap(path), attribute, value, options);
+    }
+
+    long getTotal() {
+        ensureHealth();
+        return total;
+    }
+
+    void ensureHealth() throws AssertionError {
+        boolean timerIsAlive = timerThread.isAlive();
+        Throwable cause = error.get();
+        if (cause != null || !timerIsAlive) {
+            throw new AssertionError("The quota aware filesystem has failed", cause);
+        }
+    }
+
+    long getRemaining() {
+        ensureHealth();
+        return remaining;
+    }
+
+    @Override
+    public FileChannel newFileChannel(Path path, Set<? extends OpenOption> options, FileAttribute<?>... attrs) throws IOException {
+        return delegate.newFileChannel(QuotaAwarePath.unwrap(path), options, attrs);
+    }
+
+    @Override
+    public AsynchronousFileChannel newAsynchronousFileChannel(
+        Path path,
+        Set<? extends OpenOption> options,
+        ExecutorService executor,
+        FileAttribute<?>... attrs
+    ) throws IOException {
+        return delegate.newAsynchronousFileChannel(QuotaAwarePath.unwrap(path), options, executor, attrs);
+    }
+
+    /**
+     * Normally only used in testing. Avoids thread leak when life cycle of this
+     * object doesn't follow that of the JVM.
+     *
+     * @throws IOException if something goes wrong
+     */
+    @Override
+    public void close() throws IOException {
+        if (closed.compareAndSet(false, true)) {
+            timer.cancel();
+
+            // If there was a currently executing task wait for it, to
+            // avoid false positives on file handle leak.
+            try {
+                timerThread.join();
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt(); // Restore interrupted flag
+            }
+
+            try {
+                // Inlined org.apache.lucene.util.IOUtils.close(...) to
```

Comment:
Now that this is becoming a plugin and as such is released for every ES version, it could just use the Lucene version in ES

Reply:
Ah, I meant to do that, thanks Konrad.